README.md
# CSE230-Project

## Rhythm Dash: The Haskell Rhythm Game

### Project Collaborators

| Name             | GitHub Account         |
|------------------|------------------------|
| Yimeng Wang      | yyyyyymmm              |
| Luyao Ma         | luyaomacs              |
| Ruoyang Li       | koichi-domoto          |
| Kerong Xiang     | chloe020624@gmail.com  |

### Introduction
Rhythm Dash is a rhythm game built with Haskell and the brick library. The reference game is MuseDash.

![mushdash](./img/MushDash.png)

### Gameplay
As the background music plays, notes aligned with the rhythm will flow from the right towards the judgment points on the left. Players interact by pressing corresponding keys at the precise moment to get a high score. Additionally, Players must maintain their blood level, which depletes when missing hitting. The game incorporates special items, like the "Double" power-up, which doubles the score for a limited time (we call it "bonus time"). And the blood-level-restoring items, which replenish the player's blood level. Players need to keep their blood level from emptying, or else they will lose the game.

### Core Features
- Synced notes with background music.
- Special in-game items to boost scores and replenish blood level.
- Real-time scoring that reflects the accuracy of player interactions.

### Goals
- Display the start screen and game menus.
- Load and play background music.
- Display flowing notes that aligned with the rhythm.
- Implement user interface with multiple keys for hitting notes and getting special items.
- Implement real-time scoring system with immediate feedback.
- Show final scores and other performance metrics in the result page.

### Further Goals
- Introduce varied note types, such as short and long presses, with combo sequences.
- Incorporate different difficulty levels and song options.
- Develop local multiplayer mode or networked multiplayer mode.

### Implementation Roadmap
- UI Development: Use the brick library to create the start screen, note rows, note graphics, current score display, hit feedback, and the results page.
- Game Logic: Integrate music playback, note generation in sync with the rhythm, and score computation based on player actions.
- Testing: Perform unit testing for each component to ensure reliability and performance.
- Extended Features: Implement further goals, refine features, and conduct comprehensive testing.
- Finalization: Complete the project with thorough documentation and prepare for the final presentation and demonstration.

## Updates

### Architecture of Application

- `Sound`: Engaging background music that accompanies the game-play experience.
- `Graphics`: Dynamic notes appearing in sync with the rhythm.
- `Interaction`: The player's score dynamically updates as they press corresponding keys on the keyboard.

### Challenges and Solutions

So far, we have set up the basic framework of the whole program. With the background music playing, notes that are synchronized with the rhythm flow from the right to the judgment point on the left. Players strive to achieve high scores by accurately pressing the corresponding keys at the right moment.

Though the challenge is that we need to accelerate the pace of our progress, we still think we are able to achieve the basic functionality we envisioned earlier. In the upcoming time, we will continue working on completing the content of our project to achieve our goals.

Next step, we need to :

1. Showcase flowing notes that are synchronized with the rhythm.
2. Develop a user interface with multiple keys for hitting notes, acquiring items, and avoiding traps.
3. Implement a real-time scoring system that provides immediate feedback.
4. Display final scores and other performance metrics on the results page.

### Do we expect to meet our goals until the deadline?

- Yes

## Game Logic Implementation

```
.
|-- app
|   `-- Main.hs
|-- assets
|   |-- MusicChoice.txt
|   `-- bestResult.txt
|-- src
|   |-- GameUI.hs
|   |-- GameUtils.hs
|   |-- MusicUI.hs
|   `-- StartUI.hs
`-- test
    |-- Spec.hs
    `-- TestUtil.hs
```


### Start Menu

The implementation of start menu is in [StartUI.hs](https://github.com/yyyyyymmm/CSE230-Project/blob/main/src/StartUI.hs).

![startmenu](./img/startmenu.png)

### Choose Music

The implementation of choosing music is in [MusicUI.hs](https://github.com/yyyyyymmm/CSE230-Project/blob/main/src/MusicUI.hs).
![startmenu](./img/choosemusic.png)

### Game Playing

The main part of the game UI is in [GameUI.hs](https://github.com/yyyyyymmm/CSE230-Project/blob/main/src/GameUI.hs), underlying game logit is in [GameUtils.hs](https://github.com/yyyyyymmm/CSE230-Project/blob/main/src/GameUtils.hs).
![startmenu](./img/GameUI.png)

## Source Acknowledgment
We learned the basic framework of how to define App, how to set up a channel to handle Tick events and how to handle keyboard events from the implementation of a dino game at [https://github.com/arcticmatt/dino-brick/tree/dino](https://github.com/arcticmatt/dino-brick/tree/dino) and the implementation of a snake game at [https://github.com/samtay/tetris/tree/master](https://github.com/samtay/tetris/tree/master). We also learned the basic commands to draw UI. However, all the game logic and the UI are implemented by ourselves.

We discovered an existing implementation of another rhythm game at [https://github.com/erichanxin/CSE-230-Project-Rhythm-Game](https://github.com/erichanxin/CSE-230-Project-Rhythm-Game). The rhythm game we implemented is an imitation for the game MuseDash, while their rhythm game is similar to the game Rhythm Master. There are some common things between the rhythm games but more differences. During the implementation, we were inspired by how they recorded the notes and the shift of the time and position of notes. We referred to these ideas but made our own implementation. The vast majority of the code is still developed by ourselves.

### Feature Extensions

1. Start Menu: Designed the start interface with Start Game, Choose Music and Quit options.

2. Music Selection: Designed the music selection interface. This interface allows players to select from different music options.

3. Music Playing: Utilized ‘mpv’ to implement music playing which can be used both on Linux and MacOS.

4. Notes Movement: Developed notes moving from left to right according to the beat of music. The notes will disappear after reaching the hit points.

5. Game Character: Designed a boxman moving up and down according to the key press during the rhythm game. The position of the boxman indicates which line is hit.

6. Hit Judgement and Score: Developed the notes hit judgment system with three different levels: Perfect for 2 points, Good for 1 point, Miss for 0 point.

7. HP System: Designed the hp system to record the HP during the game. The game ends when HP is 0.

8. Prop System: Designed the prop system. The props have different effects of healing or entering bonus time. The healing will +2 HP. The bonus time will earn x2 scores.

9. Best Results: Implemented a feature to display the best results achieved by the player. This includes high scores and max combos.









CHANGELOG.md
# Changelog for `rhythmdash`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/Spec.hs
import TestUtil
import GameUtils
import Test.QuickCheck

main :: IO ()
main = do
    putStrLn "blood decrease 1 after Miss"
    quickCheck prop_evaluateBlood
    putStrLn "bonusTime counter"
    quickCheck prop_BonusTime
    putStrLn "number of notes decreases after move function"
    quickCheck prop_notes_size
    putStrLn "all notes are positive"
    quickCheck prop_notes_positive
    putStrLn "note value minus 1 after falling"
    quickCheck prop_notes_one_unit
    putStrLn "comboCount after update"
    quickCheck prop_combo_counter_update
    putStrLn "comboCount after hit"
    quickCheck prop_combo_counter_hit


test/TestUtil.hs
module TestUtil where
import Test.QuickCheck
import Control.Monad (liftM2)
import Test.QuickCheck.Monadic (monadicIO, run, assert)
import GameUtils
import Brick (runEventM)
instance Arbitrary HitState where
  arbitrary = do
    x <- elements [Perfect, Good, Miss]
    return x

instance Arbitrary Process where
  arbitrary = do
    x <- elements [NoneHit , Hit ]
    return x

instance Arbitrary Key where
  arbitrary = do
    x <- elements [KeyW , KeyS , KeyJ , KeyI]
    return x

newtype BonusTime = BonusTime Int deriving Show

instance Arbitrary BonusTime where
    arbitrary = BonusTime <$> choose (0, 10)


prop_evaluateBlood:: Int -> Process -> Key -> HitState -> Bool
prop_evaluateBlood blood e k s 
    | newblood == blood - 1 = s == Miss && e == NoneHit
    | newblood == blood + 2 = s == Perfect && e == Hit && (k == KeyW || k == KeyI)
    | otherwise = newblood == blood
    where 
        newblood = evaluateBlood blood e k s


prop_BonusTime :: BonusTime -> Process -> Key -> HitState -> Bool
prop_BonusTime (BonusTime t) e k s
  | newT == 10 = e == Hit && (k == KeyI || k == KeyW) && s == Good || ( t == 10 && e == Hit)
  | newT == t - 1 = t > 0 
  | otherwise = newT == t && e == Hit || t == 0 
  where
      newT = evaluateBonusTime t e k s

-- move
prop_notes_size :: [[Int]] -> Property
prop_notes_size s = length s == 4 ==> all (\(a, b) -> length a <= length b) (zip newS s)
  where newS = move s

prop_notes_positive :: [[Int]] -> Bool
prop_notes_positive = all (>0) . concat . move

prop_notes_one_unit :: [[Int]] -> Bool
prop_notes_one_unit s = all (`elem` (concat s)) (map (\h -> h+1) (concat (move s)))


-- evaluateCombo
-- 反向核对 NoneHit 不改变compo计数 
prop_combo_counter_update :: Int -> HitState -> Bool
prop_combo_counter_update c hs = case evaluateCombo c NoneHit hs of
    0       -> hs == Miss || c == 0
    newC    -> newC == c

-- Hit counter ++
prop_combo_counter_hit :: Int -> HitState -> Bool
prop_combo_counter_hit c hs = case evaluateCombo c Hit hs of
    0       -> hs == Miss || (hs /= Miss && c == -1)
    newC    -> newC == c+1



src/GameUI.hs
module GameUI where

import GameUtils

import System.Directory
import System.FilePath ((</>))
import System.FilePath.Windows (FilePath)
import System.IO
import Control.Monad.IO.Class (liftIO)
import Control.Concurrent (threadDelay)
import Brick
  ( App(..)
  , BrickEvent(..)
  , EventM
  , Next
  , AttrMap
  , AttrName
  , Widget
  , continue
  , halt
  , str
  , withBorderStyle
  , attrMap
  , attrName
  , withAttr
  , emptyWidget
  , hLimit
  , vLimit
  , vBox
  , hBox
  , on
  , customMain
  , neverShowCursor
  , padAll
  , padTop
  , padBottom
  , Padding(..)
  , fg
  , (<+>)
  )
import Brick.BChan
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Border.Style as BS
import qualified Brick.Widgets.Center as C
import qualified Brick.Util as U
import qualified Graphics.Vty as V
import Control.Concurrent (forkIO, threadDelay)
import Control.Exception (bracket)
import Control.Monad (forever, unless, void, when)

data Tick = Tick

app :: App Game Tick ()
app = App { appDraw = drawUI
          , appChooseCursor = neverShowCursor
          , appHandleEvent = appEvent
          , appStartEvent = return
          , appAttrMap = const theMap
          }

startGame :: IO ()
startGame = do
  chan <- newBChan 10
  void $ forkIO $ forever $ writeBChan chan Tick >> threadDelay 100000
  g <- initGame
  let builder = V.mkVty V.defaultConfig
  initialVty <- builder
  void $ customMain initialVty builder (Just chan) app g

appEvent :: Game -> BrickEvent () Tick -> EventM () (Next Game)
appEvent g (AppEvent Tick) = update g
appEvent g (VtyEvent (V.EvKey key [])) = keyPress key g
appEvent g _ = continue g

keyPress :: V.Key -> Game -> EventM () (Next Game)
keyPress key g = case key of
  V.KChar 's' -> continue $ hit KeyS g
  V.KChar 'S' -> continue $ hit KeyS g
  V.KChar 'j' -> continue $ hit KeyJ g
  V.KChar 'J' -> continue $ hit KeyJ g
  V.KChar 'w' -> continue $ hitTool KeyW g
  V.KChar 'W' -> continue $ hitTool KeyW g
  V.KChar 'i' -> continue $ hitTool KeyI g
  V.KChar 'I' -> continue $ hitTool KeyI g
  V.KChar 'r' -> restartGame g
  V.KChar 'R' -> restartGame g
  V.KChar 'q' -> quit g
  V.KChar 'Q' -> quit g
  _ -> continue $ g

quit :: Game -> EventM () (Next Game)
quit g = do
  liftIO $ stopMusic (_music g)
  liftIO $ threadDelay 200000
  halt g

theMap :: AttrMap
theMap = attrMap V.defAttr
  [ (downAttr, V.white `on` V.rgbColor 217 209 155)
  , (upAttr, V.white `on` V.rgbColor 128 137 122)
  , (vLineAttr, V.withForeColor V.defAttr V.brightCyan )
  , (hLineAttr, V.withForeColor V.defAttr V.brightCyan)
  , (bonusTimeAttr, fg V.red `V.withStyle` V.bold)
  , (scoreAttr, V.white `on` V.rgbColor 150 80 75)
  , (quitAttr, V.white `on` V.rgbColor 217 209 155)
  , (comboAttr, V.white `on` V.rgbColor 128 137 122)
  , (eyeAttr, V.withForeColor V.defAttr V.brightCyan)
  , (propAttr, V.red `on` V.red)
  , (perfectAttr, fg V.red `V.withStyle` V.bold)
  , (goodAttr, fg V.magenta)
  ]

propAttr :: AttrName
propAttr = attrName "prop"

emptyAttr :: AttrName
emptyAttr = attrName "empty"

upAttr :: AttrName
upAttr = attrName "upAttr"

downAttr :: AttrName
downAttr = attrName "downAttr"

vLineAttr :: AttrName
vLineAttr = attrName "vLineAttr"

hLineAttr :: AttrName
hLineAttr = attrName "hLineAttr"

bonusTimeAttr :: AttrName
bonusTimeAttr = attrName "bonusTimeAttr"

perfectAttr :: AttrName
perfectAttr = attrName "perfect"

goodAttr :: AttrName
goodAttr = attrName "good"

scoreAttr :: AttrName
scoreAttr = attrName "score"

quitAttr :: AttrName
quitAttr = attrName "quit"

comboAttr :: AttrName
comboAttr = attrName "combo"

eyeAttr :: AttrName
eyeAttr = attrName "eye"
drawUI :: Game -> [Widget ()]
drawUI g =
  if (_end g) || ((_blood g) <= 0) then 
    [C.vCenter $ C.hCenter $ drawGameOver g] 
  else (
    [ C.vCenter $ hBox $ 
      [vBox $ [ drawScore (_score g, _combo g, _blood g), padAll 1 (str " "), drawHit (_hit g) ]
        , drawBlock g
        , drawNotes g
        , vBox $ [drawGuide, padTop (Pad 1) $ drawBonusTime (_bonusTime g)]
      ]
    ]
  )

drawNotes :: Game -> Widget ()
drawNotes g = withBorderStyle BS.unicodeBold
  $ B.borderWithLabel (str " Rhythm Dash ")
  $ vBox $ map hBox grid
  where
    grid = [[drawGridAt x y | x <- [0..17]] | y <- [19, 18..0]]
    drawGridAt x y
      | y >= 6 && y < 9 && x `elem` (_notes g) !! 1 = withAttr downAttr (str "     ")
      | y >= 11 && y < 14 && x `elem` (_notes g) !! 2 = withAttr upAttr (str "     ")
      | y >= 1 && y < 4 && x `elem` (_notes g)!! 0 = withAttr propAttr (str " ♥♥♥ ")
      | y >= 16 && y < 19 && x `elem` (_notes g)!! 3= withAttr propAttr (str " ♥♥♥ ")
      | x == 0 && ((y >= 6 && y < 9) || (y >= 11 && y < 14)|| (y >= 1 && y < 4) || (y >= 16 && y < 19)) = withAttr vLineAttr (str "|")
      | x == 2 && ((y >= 6 && y < 9) || (y >= 11 && y < 14)|| (y >= 1 && y < 4) || (y >= 16 && y < 19)) = withAttr vLineAttr (str "|")
      | otherwise = withAttr emptyAttr (str "     ")



drawGameOver :: Game -> Widget ()
drawGameOver g = withBorderStyle BS.unicodeBold
  $ hLimit 100
  $ B.borderWithLabel (str " Game over ")
  $ vBox $ [withAttr scoreAttr $ C.hCenter $ str ("     Final score: "++ (show $ _score g))
  , withAttr comboAttr $ C.hCenter $ str ("     Maximum combo: "++ (show $ _comboMax g))
  , withAttr scoreAttr $ C.hCenter $ str ("     History best score: "++ (show $ ((_bestResult g)!!0)))
  , withAttr comboAttr $ C.hCenter $ str ("     History best combo: "++ (show $ ((_bestResult g)!!1)))
  , withAttr quitAttr $ C.hCenter $ str ("     Press Q to go back to main page or press R to restrart.     ")
  ]

drawGuide :: Widget ()
drawGuide = withBorderStyle BS.unicodeBold
  $ B.borderWithLabel (str " Guidance ")
  $ hLimit 25
  $ vLimit 30
  $ padAll 1
  $ vBox $ [
            C.hCenter $ str $ "Hit Line 1   I", padTop (Pad 1) (str " ")
          , C.hCenter $ str $ "Hit Line 2   J", padTop (Pad 1) (str " ")
          , C.hCenter $ str $ "Hit Line 3   S", padTop (Pad 1) (str " ")
          , C.hCenter $ str $ "Hit Line 4   W", padTop (Pad 1) (str " ")
          , C.hCenter $ str $ "Restart      R", padTop (Pad 1) (str " ")
          , C.hCenter $ str $ "Quit         Q"]


drawScore :: (Int, Int, Int) -> Widget ()
drawScore (score, combo, blood) = withBorderStyle BS.unicodeBold 
  $ B.borderWithLabel (str " ScoreBoard ")
  $ hLimit 15
  $ vLimit 30
  $ padAll 1
  $ vBox $ [C.hCenter $ str $ ("Score : " ++ show score), padTop (Pad 1) (str " ")
          , C.hCenter $ str $ ("Combo : " ++ show combo), padTop (Pad 1) (str " ")
          , C.hCenter $ str $ ("Blood : " ++ show blood)]
  -- $ vBox $ [C.hCenter $ str $ show s]

drawHit :: HitState -> Widget ()
drawHit s = withBorderStyle BS.unicodeBold
  $ B.borderWithLabel (str " Hit ")
  $ hLimit 15
  $ vLimit 15
  $ padAll 1
  $ vBox $ [C.hCenter $ case s of
        Good     -> withAttr goodAttr $ str $ show s
        Perfect  -> withAttr perfectAttr $ str $ show s
        _          -> str $ show s]

drawBonusTime :: Int -> Widget ()
drawBonusTime t = withBorderStyle BS.unicodeBold
  $ B.borderWithLabel (str " Bonus Time ")
  $ hLimit 25
  $ vLimit 15
  $ C.hCenter
  $ padAll 1
  $ if t > 0 then
      withAttr bonusTimeAttr $ str $ show $ t
  else
      str $ show $ t

drawBlock :: Game -> Widget ()
drawBlock g = withBorderStyle BS.unicodeBold
  $ B.borderWithLabel (str " Boxman ")
  $ vBox rows
  where
    rows         = [hBox $ cellsInRow r | r <- [19, 18..0]]
    cellsInRow y = [(drawCoord x y) | x <- [0..2]]
    drawCoord x y 
        | x == 0 && (y >= 1 && y < 4) && (_lineNumber g) == 0 = withAttr vLineAttr (str "|")
        | x == 2 && (y >= 1 && y < 4) && (_lineNumber g) == 0 = withAttr vLineAttr (str "|")
        | x == 0 && (y >= 6 && y < 9) && (_lineNumber g) == 1 = withAttr vLineAttr (str "|")
        | x == 2 && (y >= 6 && y < 9) && (_lineNumber g) == 1 = withAttr vLineAttr (str "|")
        | x == 0 && (y >= 11 && y < 14) && (_lineNumber g) == 2 = withAttr vLineAttr (str "|")
        | x == 2 && (y >= 11 && y < 14) && (_lineNumber g) == 2 = withAttr vLineAttr (str "|")
        | x == 0 && (y >= 16 && y < 19) && (_lineNumber g) == 3 = withAttr vLineAttr (str "|")
        | x == 2 && (y >= 16 && y < 19) && (_lineNumber g) == 3 = withAttr vLineAttr (str "|")
        | x == 1 && y == 1 && (_lineNumber g) == 0 = withAttr hLineAttr (str "_____")
        | (x == 0 || x == 1) && y == 4 && (_lineNumber g) == 0 = withAttr hLineAttr (str "_____")
        | x == 1 && y == 6 && (_lineNumber g) == 1 = withAttr hLineAttr (str "_____")
        | (x == 0 || x == 1) && y == 9 && (_lineNumber g) == 1 = withAttr hLineAttr (str "_____")
        | x == 1 && y == 11 && (_lineNumber g) == 2 = withAttr hLineAttr (str "_____")
        | (x == 0 || x == 1) && y == 14 && (_lineNumber g) == 2 = withAttr hLineAttr (str "_____")
        | x == 1 && y == 16 && (_lineNumber g) == 3 = withAttr hLineAttr (str "_____")
        | (x == 0 || x == 1) && y == 19 && (_lineNumber g) == 3 = withAttr hLineAttr (str "_____")
        | x == 1 && y == 2 && (_lineNumber g) == 0 = withAttr eyeAttr (str " OvO ")
        | x == 1 && y == 7 && (_lineNumber g) == 1 = withAttr eyeAttr (str " OvO ")
        | x == 1 && y == 12 && (_lineNumber g) == 2 = withAttr eyeAttr (str " OvO ")
        | x == 1 && y == 17 && (_lineNumber g) == 3 = withAttr eyeAttr (str " OvO ")
        | otherwise             = withAttr emptyAttr (str "     ")




src/MusicUI.hs
module MusicUI where

import System.Exit (exitSuccess)

import Brick
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Border.Style as BS
import qualified Brick.Widgets.Center as C
import qualified Graphics.Vty as V
import Control.Monad.IO.Class (liftIO)

import Brick
  ( App(..), AttrMap, BrickEvent(..), EventM, Next, Widget
  , customMain, neverShowCursor, attrName, simpleMain
  , continue, halt
  , hLimit, vLimit, vBox, hBox
  , padRight, padLeft, padTop, padAll, padBottom, Padding(..)
  , withBorderStyle
  , str
  , attrMap, withAttr, emptyWidget, AttrName, on, fg
  , (<+>)
  )
import Brick.BChan (newBChan, writeBChan)
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Border.Style as BS
import qualified Brick.Widgets.Center as C
import qualified Brick.Util as U
import qualified Graphics.Vty as V
--import Data.Sequence (Seq)
--import qualified Data.Sequence as S

titleAttr :: AttrName
titleAttr = attrName "title"
startAttr :: AttrName
startAttr = attrName "start"
musicAttr :: AttrName
musicAttr = attrName "music"
quitAttr :: AttrName
quitAttr = attrName "quit"

myBorderStyle :: BS.BorderStyle
myBorderStyle = BS.BorderStyle { BS.bsCornerTL = '╔'
                                , BS.bsCornerTR = '╗'
                                , BS.bsCornerBL = '╚'
                                , BS.bsCornerBR = '╝'
                                , BS.bsIntersectFull = '╬'
                                , BS.bsIntersectL = '╠'
                                , BS.bsIntersectR = '╣'
                                , BS.bsIntersectT = '╦'
                                , BS.bsIntersectB = '╩'
                                , BS.bsHorizontal = '═'
                                , BS.bsVertical = '║'
                                }

ui :: Widget ()
ui =
  C.center
    $ vBox [ C.hCenter $ withAttr titleAttr $ str "!! Rhythm Dash !!"
           , C.hCenter $ str "^o^"
           , C.hCenter $ withBorderStyle myBorderStyle $ B.borderWithLabel (str " Choose Music ") $ hLimit 80
          --  , padTop (Pad 2) $ vBox
           $ vBox
           [ withAttr startAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Music 1"
                  , C.hCenter $ str "Press <1>"
                  ]
           , withAttr musicAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Music 2"
                  , C.hCenter $ str "Press <2>"
                  ]
           , withAttr startAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Music 3"
                  , C.hCenter $ str "Press <3>"
                  ]
           , withAttr quitAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Quit"
                  , C.hCenter $ str "Press <q>"
                  ]
           ]
    ]

theMap :: AttrMap
theMap = attrMap V.defAttr
    [ (titleAttr, V.white `on` V.blue)
    , (startAttr, V.white `on` V.rgbColor 150 80 75)
    , (musicAttr, V.white `on` V.rgbColor 128 137 122)
    , (quitAttr, V.white `on` V.rgbColor 217 209 155)
    ]

app :: App (Maybe Int) e ()
app = App
  { appDraw         = const [ui]  -- draw what "ui" defines
  , appHandleEvent  = handleEvent -- handle keyboard events
  , appStartEvent   = return      -- nothing
  -- , appAttrMap      = const $ attrMap V.defAttr [] -- default attributes
  , appAttrMap      = const theMap
  , appChooseCursor = neverShowCursor -- never show cursor
  }

handleEvent :: Maybe Int -> BrickEvent () e -> EventM () (Next (Maybe Int))
handleEvent n (VtyEvent (V.EvKey V.KEsc        _)) = halt $ Just 0
handleEvent n (VtyEvent (V.EvKey (V.KChar 'q') _)) = halt $ Just 0
handleEvent n (VtyEvent (V.EvKey (V.KChar 'Q') _)) = halt $ Just 0
handleEvent n (VtyEvent (V.EvKey (V.KChar '1') [])) = do
    liftIO $ writeMusicChoice 1
    halt $ Just 1
handleEvent n (VtyEvent (V.EvKey (V.KChar '2') [])) = do
    liftIO $ writeMusicChoice 2
    halt $ Just 2
handleEvent n (VtyEvent (V.EvKey (V.KChar '3') [])) = do
    liftIO $ writeMusicChoice 3
    halt $ Just 3
handleEvent n _ = continue n

writeMusicChoice :: Int -> IO ()
writeMusicChoice n = do
  writeFile "./assets/MusicChoice.txt" (show n)

chooseMusic :: IO Int
chooseMusic = do
  mode <- defaultMain app Nothing
  case mode of
    Nothing -> exitSuccess
    Just n -> return n


src/StartUI.hs
module StartUI where

import Brick
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Center as C
import qualified Brick.Widgets.Border.Style as BS
import qualified Graphics.Vty as V
import System.FilePath ((</>))
import System.Exit (exitSuccess)

startMenu :: IO Int
-- startMenu = defaultMain app Nothing >>= maybe exitSuccess return
startMenu = do
    musicChoice <- readchooseMusic ("./assets" </> "MusicChoice.txt")
    mode <- defaultMain (app musicChoice) Nothing
    case mode of
        Just 1 -> return 1
        Just 2 -> return 2
        _ -> exitSuccess
readchooseMusic :: FilePath -> IO Int
readchooseMusic path = do
  musicInt <- readFile path 
  return $ read musicInt

app :: Int -> App (Maybe Int) e ()
app musicChoice = App
  { appDraw         = const [ui musicChoice]
  , appHandleEvent  = handleEvent
  , appStartEvent   = return
  , appAttrMap      = const theMap
  , appChooseCursor = neverShowCursor
  }

ui :: Int -> Widget ()
ui musicChoice =
  C.center
    $ vBox [ C.hCenter $ withAttr titleAttr $ str "!! Rhythm Dash !!"
           , C.hCenter $ str "^o^"
           , C.hCenter $ withBorderStyle myBorderStyle $ B.borderWithLabel (str " Menu ") $ hLimit 80
           $ vBox
           [ withAttr startAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Start Game"
                  , C.hCenter $ str "Press <1>"
                  ]
           , withAttr musicAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Choose Music"
                  , C.hCenter $ str "Press <2>"
                  ]
           , withAttr quitAttr $ hBox [ C.hCenter $ padRight (Pad 4) $ str "Quit"
                  , C.hCenter $ str "Press <q>"
                  ]
           ]
           , C.hCenter $ str ("Current Music: " ++ show musicChoice)
    ]

handleEvent :: Maybe Int -> BrickEvent () e -> EventM () (Next (Maybe Int))
handleEvent _ (VtyEvent (V.EvKey V.KEsc        _)) = halt Nothing
handleEvent _ (VtyEvent (V.EvKey (V.KChar 'q') _)) = halt Nothing
handleEvent _ (VtyEvent (V.EvKey (V.KChar 'Q') _)) = halt Nothing
handleEvent _ (VtyEvent (V.EvKey (V.KChar '1') [])) = halt $ Just 1
handleEvent n (VtyEvent (V.EvKey (V.KChar '2') [])) = halt $ Just 2
handleEvent _ _ = continue Nothing

theMap :: AttrMap
theMap = attrMap V.defAttr
    [ (titleAttr, V.white `on` V.blue)
    , (startAttr, V.white `on` V.rgbColor 150 80 75)
    , (quitAttr, V.white `on` V.rgbColor 217 209 155)
    , (musicAttr, V.white `on` V.rgbColor 128 137 122)
    ]

titleAttr :: AttrName
titleAttr = attrName "title"
startAttr :: AttrName
startAttr = attrName "start"
quitAttr :: AttrName
quitAttr = attrName "quit"
musicAttr :: AttrName
musicAttr = attrName "music"
myBorderStyle :: BS.BorderStyle
myBorderStyle = BS.BorderStyle { BS.bsCornerTL = '╔'
                                , BS.bsCornerTR = '╗'
                                , BS.bsCornerBL = '╚'
                                , BS.bsCornerBR = '╝'
                                , BS.bsIntersectFull = '╬'
                                , BS.bsIntersectL = '╠'
                                , BS.bsIntersectR = '╣'
                                , BS.bsIntersectT = '╦'
                                , BS.bsIntersectB = '╩'
                                , BS.bsHorizontal = '═'
                                , BS.bsVertical = '║'
                                }
